apps/src/blockly/addons/cdoUtils.js

```diff
@@ -106,9 +109,168 @@ export function getUserTheme(themeOption) {
+    Math.random() <= 0.05
+  ) {
+    testJsonSerialization(workspace);
+  }
   return Blockly.Xml.domToText(Blockly.Xml.blockSpaceToDom(workspace));
-  // After supporting JSON block sources, change to:
-  // return JSON.stringify(Blockly.serialization.workspaces.save(workspace));
+}
+
+function testJsonSerialization(workspace) {
+  const FIREHOSE_STUDY = 'blockly-json';
+  const FIREHOSE_EVENT_PASS = 'blocks-identical';
+  const FIREHOSE_EVENT_FAIL = 'differences-found';
+  Blockly.Events.disable();
+
+  // Create an array of blocks based on JSON serialization of the current workspace.
+  const blockJson = Blockly.serialization.workspaces.save(workspace);
+  const tempJsonWorkspace = new Blockly.Workspace();
+  Blockly.serialization.workspaces.load(blockJson, tempJsonWorkspace);
+  const jsonBlocks = tempJsonWorkspace.getAllBlocks();
+
+  // Create an array of blocks based on the XML encoding of the current workspace.
+  const blockXml = Blockly.Xml.blockSpaceToDom(workspace);
+  const tempXmlWorkspace = new Blockly.Workspace();
+  Blockly.Xml.domToWorkspace(blockXml, tempXmlWorkspace);
+  const xmlBlocks = tempXmlWorkspace.getAllBlocks();
+
+  // compareBlockArrays returns an array of differences found.
+  const differences = compareBlockArrays(xmlBlocks, jsonBlocks);
+  const result =
+    differences.length > 0 ? FIREHOSE_EVENT_FAIL : FIREHOSE_EVENT_PASS;
+  // Log a record to Firehose/Redshift.
+  const recordData = {
+    projectUrl: dashboard.project.getShareUrl(),
+    blocklyVersion: Blockly.blockly_.VERSION,
+    differences: differences,
+  };
+  firehoseClient.putRecord(
+    {
+      study: FIREHOSE_STUDY,
+      event: result,
+      data_json: JSON.stringify(recordData),
+    },
+    {
+      alwaysPut: true, // Allows logging from development environments
+      includeUserId: true,
+    }
+  );
+  if (experiments.isEnabled(experiments.BLOCKLY_JSON)) {
+    // Avoid logging to the console for typical users.
+    console.log(recordData);
+  }
+}
+
+// Used to find differences between blocks created from xml and json sources.
+// Called when the app is initialized and each time the app runs.
+// Compares two block arrays and returns an array of differences between them.
+export function compareBlockArrays(xmlBlocks, jsonBlocks) {
+  // compareValues() will be called recursively so we need to keep track of objects
+  // that it has already compared, to prevent infinite loops caused by circular
+  // references.
+  const visited = new Set();
+  // Some block properties are expected to hold different values, or refer to blocks already in the array.
+  const keysToSkip = [
+    'parentBlock_', // a circular reference to another block in the array
+    'sourceBlock_', // a circular reference to another block in the array
+    'workspace', // a reference to the unique temporary workspaces used to create the two block arrays
+  ];
+
+  function compareValues(arg1, arg2, path = '', differences = []) {
```

Comment:
Can/should we include a test suite for `compareValues` and `compareBlockArrays` with this PR? It seems like there are a lot of well-thought-out edge cases here.

Reply:
+1 for adding unit tests

Reply:
Is there no existing solution out there that will compare objects (and handle circular references)?

Reply:
There very well could be. We also need to exclude certain properties from the comparison (e.g. `workspace`), and we want to be able to log the specific differences that are found so that we can effectively debug. For example, if the checker had an early return at the first difference found, we would potentially miss that a whole bunch of differences could be present for a given workspace. For these reasons, it seemed best just to implement my own utility. Also, I'm planning to add unit tests in the morning that should help prove that this utility does what it's meant to. The test will include cases for the exceptions we care about.

Reply:
Unit tests have been added.

Reply:
One thought: Do we need unit tests? (sorry, I know I'm late to the game here)

I usually think of unit tests as something that keeps us from causing regressions when we change code in the future. But if I'm reading this right, this util is intended to be used only for a short period of time. We'll use it to run some tests, but then it will be deleted. It's similar to a one-off script. Because of that, we won't have future maintainers, and we won't have a risk of future regression on this code.

Reply:
> I usually think of unit tests as something that keeps us from causing regressions when we change code in the future.

I think there's another category of unit tests which are built at the same time (or even before) the code (classically in [TDD](https://en.wikipedia.org/wiki/Test-driven_development)), which are used to help in the development of code and increase confidence that it works, even if it's only temporary.  So for that reason, it seems good to have some here.

Reply:
I'm inclined to agree with @breville in this. Given that the new utility is so complex, adding unit tests are helpful right now as they give confidence that it does what it is designed to do. Planning to leave the tests in as of right now.